diff -Naur uboot_orgin/cmd/otp.c uboot/cmd/otp.c
--- uboot_orgin/cmd/otp.c	2025-12-17 16:36:35.123137739 +0800
+++ uboot/cmd/otp.c	2025-12-17 16:45:05.295118429 +0800
@@ -137,7 +137,7 @@
 	int key_type;
 	int order;
 	int need_id;
-	char information[110];
+	char *information;
 };
 
 struct otp_pro_sts {
@@ -305,33 +305,6 @@
 	return rid_num;
 }
 
-static void sb_sha256(u8 *src, u32 len, u8 *digest_ret)
-{
-	sha256_context ctx;
-
-	sha256_starts(&ctx);
-	sha256_update(&ctx, src, len);
-	sha256_finish(&ctx, digest_ret);
-}
-
-static void sb_sha384(u8 *src, u32 len, u8 *digest_ret)
-{
-	sha512_context ctx;
-
-	sha384_starts(&ctx);
-	sha384_update(&ctx, src, len);
-	sha384_finish(&ctx, digest_ret);
-}
-
-static void sb_sha512(u8 *src, u32 len, u8 *digest_ret)
-{
-	sha512_context ctx;
-
-	sha512_starts(&ctx);
-	sha512_update(&ctx, src, len);
-	sha512_finish(&ctx, digest_ret);
-}
-
 static u32 chip_version(void)
 {
 	u32 revid0, revid1;
@@ -475,10 +448,10 @@
 	printf("%08X\n", buf[2]);
 	printf("%08X\n", buf[3]);
 	writel(otp_addr, OTP_ADDR); //Compare address
-	writel(buf[0], OTP_COMPARE_1); //Compare data 1
-	writel(buf[1], OTP_COMPARE_2); //Compare data 2
-	writel(buf[2], OTP_COMPARE_3); //Compare data 3
-	writel(buf[3], OTP_COMPARE_4); //Compare data 4
+	writel(~buf[0], OTP_COMPARE_1); //Compare data 1
+	writel(~buf[1], OTP_COMPARE_2); //Compare data 2
+	writel(~buf[2], OTP_COMPARE_3); //Compare data 3
+	writel(~buf[3], OTP_COMPARE_4); //Compare data 4
 	writel(0x23b1e363, OTP_COMMAND); //Compare command
 	wait_complete();
 	ret = readl(OTP_STATUS); //Compare command
@@ -1394,7 +1367,7 @@
 static void _otp_print_key(u32 header, u32 offset, u8 *data)
 {
 	const struct otpkey_type *key_info_array = info_cb.key_info;
-	struct otpkey_type key_info;
+	struct otpkey_type key_info = { .value = -1 };
 	int key_id, key_offset, key_type, key_length, exp_length;
 	int len = 0;
 	int i;
@@ -1408,7 +1381,6 @@
 	printf("\nKey[%d]:\n", offset);
 	printf("Header: %x\n", header);
 
-	key_info.value = -1;
 	for (i = 0; i < info_cb.key_info_len; i++) {
 		if (key_type == key_info_array[i].value) {
 			key_info = key_info_array[i];
@@ -1912,6 +1884,18 @@
 	return OTP_SUCCESS;
 }
 
+static void do_hash(const void *data, int data_len, const char *algo_name, uint8_t *value)
+{
+        struct hash_algo *algo;
+
+        if (hash_lookup_algo(algo_name, &algo)) {
+                debug("Unsupported hash alogrithm\n");
+                return;
+        }
+
+        algo->hash_func_ws(data, data_len, value, algo->chunk_size);
+}
+
 static int otp_verify_image(u8 *src_buf, u32 length, u8 *digest_buf, int version)
 {
 	u8 digest_ret[48];
@@ -1919,11 +1903,11 @@
 
 	switch (version) {
 	case 1:
-		sb_sha256(src_buf, length, digest_ret);
+		do_hash(src_buf, length, "sha256", digest_ret);
 		digest_len = 32;
 		break;
 	case 2:
-		sb_sha384(src_buf, length, digest_ret);
+		do_hash(src_buf, length, "sha384", digest_ret);
 		digest_len = 48;
 		break;
 	default:
@@ -2549,13 +2533,13 @@
 		printf("otp verify does not support SHA224\n");
 		return OTP_FAILURE;
 	case 256:
-		sb_sha256(sec_image, sign_image_size, digest_ret);
+		do_hash(sec_image, sign_image_size, "sha256", digest_ret);
 		break;
 	case 384:
-		sb_sha384(sec_image, sign_image_size, digest_ret);
+		do_hash(sec_image, sign_image_size, "sha384", digest_ret);
 		break;
 	case 512:
-		sb_sha512(sec_image, sign_image_size, digest_ret);
+		do_hash(sec_image, sign_image_size, "sha512", digest_ret);
 		break;
 	default:
 		printf("SHA Algorithm is invalid\n");
@@ -2642,9 +2626,9 @@
 	int i;
 	int pass = 0;
 
-	ret = uclass_get_device_by_driver(UCLASS_MOD_EXP, DM_GET_DRIVER(aspeed_acry), &mod_exp_dev);
+	ret = uclass_get_device(UCLASS_MOD_EXP, 0, &mod_exp_dev);
 	if (ret) {
-		printf("RSA engine: Can't find aspeed_acry\n");
+		printf("RSA: Can't find RSA driver\n");
 		return OTP_FAILURE;
 	}
 
@@ -3451,3 +3435,4 @@
 	   "otp verify <addr>\n"
 	   "otp invalid [o] <header_offset>\n"
 	  );
+
